indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,711 @@
+    if (requestBatch.isEmpty()) {
+      return true;
+    } else if (!isLeader.get()) {
+      requestBatch.failPendingRequests("Cannot allocate segment if not leader");
+      return true;
+    }
+
+    log.debug(
+        "Processing [%d] requests for batch [%s], queue time [%s].",
+        requestBatch.size(),
+        requestKey,
+        requestBatch.getQueueTime()
+    );
+
+    final long startTimeMillis = System.currentTimeMillis();
+    final int batchSize = requestBatch.size();
+    emitBatchMetric("task/action/batch/size", batchSize, requestKey);
+    emitBatchMetric("task/action/batch/queueTime", (startTimeMillis - requestBatch.getQueueTime()), requestKey);
+
+    final Set<DataSegment> usedSegments = retrieveUsedSegments(requestKey);
+    final int successCount = allocateSegmentsForBatch(requestBatch, usedSegments);
+
+    emitBatchMetric("task/action/batch/attempts", 1L, requestKey);
+    emitBatchMetric("task/action/batch/runTime", (System.currentTimeMillis() - startTimeMillis), requestKey);
+    log.info("Successfully processed [%d / %d] requests in batch [%s].", successCount, batchSize, requestKey);
+
+    if (requestBatch.isEmpty()) {
+      log.debug("All requests in batch [%s] have been processed.", requestKey);
+      return true;
+    }
+
+    // Requeue the batch only if used segments have changed
+    log.debug("There are [%d] failed requests in batch [%s].", requestBatch.size(), requestKey);
+    final Set<DataSegment> updatedUsedSegments = retrieveUsedSegments(requestKey);
+
+    if (updatedUsedSegments.equals(usedSegments)) {
+      requestBatch.failPendingRequests("Allocation failed probably due to conflicting segments.");
+      return true;
+    } else {
+      log.debug("Used segments have changed. Requeuing failed requests.");
+      return false;
+    }
+  }
+
+  private Set<DataSegment> retrieveUsedSegments(AllocateRequestKey key)
+  {
+    return new HashSet<>(
+        metadataStorage.retrieveUsedSegmentsForInterval(
+            key.dataSource,
+            key.preferredAllocationInterval,
+            Segments.ONLY_VISIBLE
+        )
+    );
+  }
+
+  private int allocateSegmentsForBatch(AllocateRequestBatch requestBatch, Set<DataSegment> usedSegments)
+  {
+    int successCount = 0;
+
+    // Find requests whose row interval overlaps with an existing used segment
+    final Set<SegmentAllocateRequest> allRequests = requestBatch.getRequests();
+    final Set<SegmentAllocateRequest> requestsWithNoOverlappingSegment = new HashSet<>();
+
+    if (usedSegments.isEmpty()) {
+      requestsWithNoOverlappingSegment.addAll(allRequests);
+    } else {
+      final Interval[] sortedUsedSegmentIntervals = getSortedIntervals(usedSegments);
+      final Map<Interval, List<SegmentAllocateRequest>> overlapIntervalToRequests = new HashMap<>();
+
+      for (SegmentAllocateRequest request : allRequests) {
+        // If there is an overlapping used segment, the interval of the used segment
```

Comment:
The allocation for a single segment considers overlap with queryGranularity.bucket(row.timestamp())
This code fetches all used segments overlapping with segmentGranularity.bucket(row.timestamp()). I think this could lead to a case where the row is allocated to a used segment interval to which it doesn't belong

Reply:
The used segments for the allocate interval (`segmentGranularity.bucket(row.timestamp())`) are fetched so that we don't have to make the metadata call multiple times.
The logic hasn't changed because we use the `rowInterval` (`queryGranularity.bucket(row.timestamp())`) to find the final overlapping segments. Please look at the line after this.